File: python/test/t_Indices_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()


size = 4
indices = ot.Indices(size, 0)
indices.fill()
print("indices=", repr(indices))
print("are indices valid with bound=", size, "? ", indices.check(size))
print("are indices valid with bound=", size // 2, "? ", indices.check(size // 2))
indices[0] = indices[size - 1]
print("indices after transformation=", repr(indices))
print("are indices valid with bound=", size, "? ", indices.check(size))
print("complement wrt 7=", indices.complement(7))
print("empty complement=", ot.Indices().complement(7))

# slicing
indices2 = ot.Indices((1, 2, 3, 4, 5))
print("indices2[2] = ", indices2[2])
print("indices2[1:3] = ", indices2[1:3])
print("indices2[:] = ", indices2[:])
indices2[1:3] = (88, 99)
print("indices2 = ", indices2)

index = indices2.index(88)
assert indices2[index] == 88, "wrong index"

with ott.assert_raises(TypeError):
    index = indices2.index(77777)

# Check contains
indices = ot.Indices()
assert not (0 in indices), "wrong contains"
assert not (1 in indices), "wrong contains"
indices = ot.Indices((1, 2, 3, 4, 5))
assert 1 in indices, "wrong contains"
assert 5 in indices, "wrong contains"
indices = ot.Indices((3, 5, 7))
assert 3 in indices, "wrong contains"
assert 5 in indices, "wrong contains"
assert 7 in indices, "wrong contains"
assert not (2 in indices), "wrong contains"
assert not (4 in indices), "wrong contains"
assert not (0 in indices), "wrong contains"
assert not (1 in indices), "wrong contains"

# Check norm1() and normInf()
indices = ot.Indices(())
assert indices.norm1() == 0, "wrong norm1"
assert indices.normInf() == 0, "wrong norm1"
indices = ot.Indices((3, 5, 7))
assert indices.norm1() == 15, "wrong norm1"
assert indices.normInf() == 7, "wrong norm1"

File: lib/test/t_Indices_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class Indices for standard methods
 *
 *  Copyright 2005-2026 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    UnsignedInteger size = 4;
    Indices indices(size, 0);
    indices.fill();
    fullprint << "indices=" << indices << std::endl;
    fullprint << "are indices valid with bound=" << size << "? " << (indices.check(size) ? "true" : "false") << std::endl;
    fullprint << "are indices valid with bound=" << size / 2 << "? " << (indices.check(size / 2) ? "true" : "false") << std::endl;
    indices[0] = indices[size - 1];
    fullprint << "indices after transformation=" << indices << std::endl;
    fullprint << "are indices valid with bound=" << size << "? " << (indices.check(size) ? "true" : "false") << std::endl;
    fullprint << "complement wrt 7=" << indices.complement(7) << std::endl;


    // braced-init-list
    Indices indices2 = {1, 2, 3};
    fullprint << "indices2=" << indices2 << std::endl;
    // copy-list-initialization
    indices2 = {4, 5};
    fullprint << "indices2=" << indices2 << std::endl;

    // Test contains()
    Indices indices3 = {};
    if(indices3.contains(0)) throw InvalidArgumentException(HERE) << "contains(0)";
    if(indices3.contains(1)) throw InvalidArgumentException(HERE) << "contains(1)";
    Indices indices4 = {1, 2, 3};
    if(!indices4.contains(1)) throw InvalidArgumentException(HERE) << "contains(1)";
    if(indices4.contains(4)) throw InvalidArgumentException(HERE) << "contains(4)";

    // Test normInf() and norm1()
    Indices indices6 = {};
    if(indices6.normInf() != 0.0) throw InvalidArgumentException(HERE) << "norm";
    if(indices6.norm1() != 0.0) throw InvalidArgumentException(HERE) << "norm";
    Indices indices7 = {1, 2, 3};
    if(indices7.normInf() != 3.0) throw InvalidArgumentException(HERE) << "norm";
    if(indices7.norm1() != 6.0) throw InvalidArgumentException(HERE) << "norm";

  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }

  return ExitCode::Success;
}
